ruuvitag_sensor/ble_communication.py

```diff
@@ -123,3 +129,110 @@ def get_data(mac, bt_device=''):
                 break
 
         return data
+
+
+class BleCommunicationBleson(BleCommunication):
+    '''Bluetooth LE communication with Bleson'''
+
+    @staticmethod
+    def _run_get_data_background(queue, shared_data, bt_device):
+        (observer, q) = BleCommunicationBleson.start(bt_device)
+
+        for line in BleCommunicationBleson.get_lines(q):
+            if shared_data['stop']:
+                break
+            try:
+                mac = line.address.address
+                if mac in shared_data['blacklist']:
+                    continue
+                data = line.service_data or line.mfg_data
+                if data is None:
+                    continue
+                queue.put((mac, data))
+            except GeneratorExit:
+                break
+            except:
+                continue
+
+        BleCommunicationBleson.stop(observer)
+
+    @staticmethod
+    def start(bt_device=''):
+        '''
+        Attributes:
+           device (string): BLE device (default 0)
+        '''
+
+        if not bt_device:
+            bt_device = 0
+        else:
+            # Old communication used hci0 etc.
+            bt_device = bt_device.replace('hci', '')
+
+        log.info('Start receiving broadcasts (device %s)', bt_device)
+
+        q = Queue()
+
+        adapter = get_provider().get_adapter(int(bt_device))
+        observer = Observer(adapter)
+        observer.on_advertising_data = q.put
+        observer.start()
+
+        return (observer, q)
+
+    @staticmethod
+    def stop(observer):
+        observer.stop()
+
+    @staticmethod
+    def get_lines(queue):
+        try:
+            while True:
+                next_item = queue.get(True, None)
+                yield next_item
+        except KeyboardInterrupt as ex:
+            return
+        except Exception as ex:
+            log.info(ex)
+            return
+
+    @staticmethod
+    def get_datas(blacklist=[], bt_device=''):
+        m = Manager()
+        q = m.Queue()
+
+        # Use Manager dict to share data between processes
+        shared_data = m.dict()
+        shared_data['blacklist'] = blacklist
+        shared_data['stop'] = False
+
+        # Start background process
+        proc = Process(target=BleCommunicationBleson._run_get_data_background, args=[q, shared_data, bt_device])
```

Comment:
If `get_data` or `get_datas` get called in succession, it seems this becomes a bit of a fork bomb because of that while True.

Reply:
Added Bleson as an own adapter (https://github.com/ttu/ruuvitag-sensor/pull/99) so we can merge Bleson communication to master. We can then start to fix/improve the implementation there.

Can't test that new PR before next week, so this has to wait for a while.

Reply:
Reason for using Process was that recreating Bleson connections was leaking memory. Mentioned here: https://github.com/ttu/ruuvitag-sensor/issues/78

Reply:
I see...

Let me explain my current situation:

I'm trying to get this to work on Home Assistant with Hass.io. I managed to add bluetooth.h to the base python image and it's now it's now released and supports bleson.

This branch seems to have worked reasonably for some people but it stops working after a while (after a couple of days): 
https://github.com/JonasR-/ruuvi_hass/pull/4

For me I managed to get it to work, but it seems to bork the whole setup a couple of minutes in, and I'm assuming that it's either not closing the sockets leading to some max files opened or overwhelming the container with multiple processes requesting IO.

So it seems the biggest problems are in the long term use. 

I was playing around on the weekend to try to get a simplified version of this library that maintains the connection open while keeping the latest state of the sensors in the client, and as a learning exercise: https://github.com/sergioisidoro/simple-ruuvitag

I haven't reached any decent result, unfortunately. Tell me how can I help here :)

Reply:
I did some long duration testing with BlueZ and Bleson, and based on that I added the note to README you found. Test was only around 24h and I just checked that memory consumption or process count wasn’t growing.

https://github.com/ttu/ruuvitag-sensor/blob/master/developer_notes.md#run-long-duration-tests

I think I didn’t try a test which executes `get_datas` multiple times. Maybe you could run something like this?

```py
import time
from ruuvitag_sensor.ruuvi import RuuviTagSensor, RunFlag

# RunFlag for stopping execution at desired time
run_flag = RunFlag()

def handle_data(datas):
  print(datetime.utcnow().isoformat())
  print(datas)
  run_flag.running = False

macs = []

while True:
  run_flag.running = True
  RuuviTagSensor.get_datas(handle_data, macs, run_flag)
  time.sleep(1)
```

If that leaks memory and doesn’t kill all processes correctly, maybe we could change BleCommunicationBleson to create just one process in the beginning and always use that?

Of course proper fix would be to do the fix to Bleson as it seemed that `observer.stop()` didn’t do proper cleanup. Apparently when I noticed this didn’t have enough time to inspect it or to create an issue, so don’t remember what was wrong with it anymore...

Also if you can test that the new PR (https://github.com/ttu/ruuvitag-sensor/pull/99) works so we can then merge it to master and start fixing this from there?

Reply:
Just received bluetooth dongle from mail so I can run testing environment on my laptop instead of Raspberry Pi. Integrated chip refused to co-operate with VM. And just got all set up into homeassistant. Currently running @sergioisidoro 's version of ruuvi_hass. Polling from 6 ruuvitags with default settings so any issues should escalate pretty quickly.

Reply:
It seems that now RuuviTagSensor.get_datas() is able to recover from occasional hiccups. There might be one hour without any values and logs full of 30sec timeout errors. And then suddenly data collection starts again. Older versions on Raspberry Pi would just freeze and would not recover without reboot. So I'd say this is huge improvement atm.

Just let me know if you need help testing stuff out.

Reply:
@peltsippi Thanks for testing!

Can you send me the timeout errors you got?

If you have time, Bleson adapter branch (https://github.com/ttu/ruuvitag-sensor/pull/99) needs some testing before I can merge it to the master and release a new package. Running verification tests should be enough (https://github.com/ttu/ruuvitag-sensor/blob/add-bleson-adaptor/developer_notes.md#exeuting-verification-test).

Reply:
Sorry things got really hectic suddenly for me. I got something like 30 hours uptime and then something broke in a way that homeassistant was unable to recover without reboot.

This is the error that occurs, some of them might be because I had my laptops sleep settings wrong initially and the adapter probably fell asleep. Unhandled GAP pos=13 type=0x02 len=3 is the last warning before polling froze completely.

> 2020-03-27 17:47:00 WARNING (HCISocketPoller) [bleson] TODO: Unhandled GAP type, pos=13 type=0x02 len=3
2020-03-27 17:47:00 WARNING (HCISocketPoller) [bleson] Data: 01 00 00 a9 d5 44 1a 9e a0 15 02 01 06 03 02 ee fe 0d ff 6b 00 72 08 ef e3 d9 01 00 00 00 00

Checked statistics for error messages also in case it helps somehow:

```
pos | type | len | qty
13 | 0x02 | 3 | 5759
13 | 0x02 | 5 | 466
13 | 0x0a | 2 | 4404
13 | 0x19 | 3 | 15
17 | 0x02 | 3 | 15
17 | 0x19 | 3 | 756
19 | 0x0a | 2 | 466
21 | 0x0a | 2 | 756
24 | 0x19 | 3 | 157
25 | 0x0a | 2 | 3
29 | 0x12 | 5 | 133
35 | 0x0a | 2 | 133
```

Reply:
@peltsippi this might not be related to this library but with the home assistant configuration itself (and ruuvi_hass)

I have noticed that unlike the previous adapter, bleson can issue a blocking io to the adapter, and if the resources are not freed appropriately, i would guess that you would need to manually force close the sockets / files, or reboot. Especially if you have multiple `ruuvi` platforms.

Right now ruuvi_hass polling strategy is concurrent from multiple threads. Without a semaphore this can lead to 2 hass sensors making a request to this library, which might lead to some incoherent state or deadlocks.

I have started exploring moving ruuvi_hass to home assistant async io loop, such that polling only triggers one scan (managed by semaphores). I have been using a mini fork from this library in https://github.com/ruuvi-friends/ruuvi-hass.io and it has been running stable for one full week now. Sensors sometimes fail to update (take longer than 10 secs to update), but I'm ok with that data granularity. All problems so far came from the way ruuvi_hass works, not from bleson, which has been working quite well. 

Only one note, tho. It's really annoying that when listening to advertisements in bleson, you only get the first measurement for each sensor, and you need to close the `observer` and start it again`.